### PR TITLE
BUG: Fix buffer readflag errors and small leaks

### PR DESCRIPTION
**Pull request by seberg**
Some small fixes found with valgrind (some more coming possibly).  The main fixes are round readonly buffer export for scalars, which had some bugs that I didn't notice before.

---

Needs to be backported to 1.20, 1.19.x doesn't include these things.

### CODE REVIEW

File: numpy/core/src/multiarray/_multiarray_tests.c.src [hunk 1]
```diff
@@ -677,11 +677,12 @@ create_custom_field_dtype(PyObject *NPY_UNUSED(mod), PyObject *args)
...
     }
     if (PyArray_RegisterDataType(dtype) < 0) {
-        /* Fix original type in the error_path == 2 case. */
+        /* Fix original type in the error_path == 2 case and delete it */
         Py_SET_TYPE(dtype, original_type);
+        Py_DECREF(dtype);
         return NULL;
     }
-    Py_INCREF(dtype);
+    Py_INCREF(dtype);  /* hold on to the original (leaks a reference) */
```
Comment (charris) [inline numpy/core/src/multiarray/_multiarray_tests.c.src:685]: > (leaks a reference)

?

**Comment by charris** [comment]
What justifies the read-only assumption?

**Comment by seberg** [comment]
I find writeable voids slightly strange and it seemed existing behavior, so did not bother to change. But should double check and/ir happy to correct it.

**Comment by charris** [comment]
Thanks Sebastian.

**Comment by seberg** [comment]
Yeah, it is old behaviour (at least 1.19.x as well). The comment about a reference leak is just a slightly weird test. Only the other path actually leaked memory also which just isn't necessary...

**Pull request merged**